orttraining/orttraining/python/training/onnxblock/blocks.py

```diff
@@ -47,10 +49,56 @@ def __call__(self, *args, **kwargs):
 
         output = self.build(*args, **kwargs)
 
-        onnx.checker.check_model(self.base, True)
+        if accessor._GLOBAL_ACCESSOR.has_path:
+            temp_external_path = self.temp_onnx_file_name + ".data"
+            onnx.save(
+                accessor._GLOBAL_ACCESSOR.model,
+                self.temp_onnx_file_name,
+                save_as_external_data=True,
+                all_tensors_to_one_file=True,
+                location=temp_external_path,
+            )
+
+            onnx.checker.check_model(self.temp_onnx_file_name, True)
+        else:
+            onnx.checker.check_model(self.base, True)
 
         return output
 
+    def infer_shapes_on_base(self):
+        """
+        Performs shape inference on the global model. If a path was used, then uses the
+        infer_shapes_path API to support models with external data.
+
+        Returns the shape-inferenced ModelProto.
+        """
+        if accessor._GLOBAL_ACCESSOR.has_path:
+            temp_external_path = self.temp_onnx_file_name + ".data"
+            onnx.save(
+                accessor._GLOBAL_ACCESSOR.model,
+                self.temp_onnx_file_name,
+                save_as_external_data=True,
+                all_tensors_to_one_file=True,
+                location=temp_external_path,
+            )
+
+            onnx.shape_inference.infer_shapes_path(self.temp_onnx_file_name)
+            # shape inferenced model is saved to original path
+            model = onnx.load(self.temp_onnx_file_name)
+
+            return model
+        else:
+            return onnx.shape_inference.infer_shapes(accessor._GLOBAL_ACCESSOR.model)
+
+    def release(self):
```

Comment:
Having to manually call `release` can be problematic. e.g. if there's an exception in the usage code before the call to `release` it would get missed.

Could we use     
- add `def __del__(self):` cleanup in destructor
- or use contextlib.contextmanager to automate cleanup
  -  https://stackoverflow.com/questions/13379742/right-way-to-clean-up-a-temporary-folder-in-python-class

I would lean towards the `__del__` usage as it requires less of the user. not sure which is the more pythonic approach.